### ISSUE TITLE
🪣 Design: Subscription/Notification Refactoring - Platform as Service approach

### ISSUE THREAD

**Issue by NakulManchanda** [hicommonwealth/commonwealth#3696]
## Description
**Subscription/Notification Refactoring**

**Research for current:**
- [ ] Create formal map of all routes and processes hitting these tables 
- [ ] Fix NotificationsCount 
- [ ] Denormalize Subscriptions
- [ ] Denormalize Notifications + NotificationsRead

## Product Requirement
How we can make notification extensible?
How can we manage subscriptions & notification delivery of third party apps?

## Tasks
- [x] Remodel - new schema and management scheme - extensible to introduce new Events by native or third party developer on platform
- [ ] Model existing events - to identify gaps and limitations
- [ ] Sample POC
- [ ] Finalize design and requirements
- [ ] Implementation

## Initial Re-modeling Proposal
https://github.com/hicommonwealth/commonwealth/wiki/Notifications

## Acceptance Criteria
- [ ] Extensible platform notification service, to allow different apps to register their notification
- [ ] Manage subscription for both main apps & third party apps
- [ ] Manage delivery for for both main apps & third party apps
- [ ] Support Unsupported Notification Delivery for both push/pull model

**Comment by NakulManchanda** [comment]
It never gained any traction yet, will started looking into this by end of this week. This is major focus for Database improvement plan.

**Comment by NakulManchanda** [comment]
@dillchen - just FYI we have ticket for this already
- Feel free to create a separate Design spike ticket or adjust acceptance criteria as per product vision

**Comment by CowMuon** [comment]
Pls note we now have a [Notifications Cleanup milestone](https://github.com/hicommonwealth/commonwealth/milestone/30) (thanks Timothee!)

**Comment by NakulManchanda** [comment]
Hope its well thought out. As I have no knowledge of what is planned.
This targets both performance and extensibility to third party devs 

Here is my proposal:
https://github.com/hicommonwealth/commonwealth/wiki/notifications

**Comment by timolegros** [comment]
My clean-up is very straightforward. Mainly aimed at reducing the size of notif + sub tables, cleaning up some indexes/columns names, removing weird/indirect `object_id` references, and finally removing notifications read table. The goal of the milestone is not necessarily to directly improve performance (though in some cases it is a by-product). Instead, the goal is to clean everything up so the existing system is very straightforward/simple. Once that is achieved we can hopefully have a smooth transition to a brand new notifications/subscription system that relies on an event/data pipeline rather than solely on the CW server and a single relational database.

The main thing I see you could focus on that has immediate results is the `viewSubscriptions` function. Optimizing that route/query is definitely something that needs to get looked at. Pretty much every other limitation except for socket.io is either handled by the removal of NotificationsRead (which has been requested by leads for some time) or by the new notifications system we will build. 

As for the socket.io issue I have said this multiple times but rooms are very light. They are by no means "not sustainable" considering that Redis can handle billions of keys and rooms use a few keys each. Each user already has their own Websocket connection. Looping through users is precisely what Socket.io does in the background the difference being it handles tracking what users are in each room for us. Until we reach 10-20k concurrent WebSockets this is not at all an issue.

**Comment by NakulManchanda** [comment]
why are we calling /viewSubscription on every screen? Are we subscribing users to room/channel? If yes, then we can stop doing that and push the notification when available.

Its not about socket.io handling connections.
Its about calling /viewSubscription throughout user session.

We anyways generate one notification per user for emails, so socket.io room subscription not even help us saving on fan out. 

We already know which users we need to send notification whether they are subscribed or not to room. Also, its not a chat channel, most of these events happen only once with no follow up message in same room if room are based on object_ids

we never had a chance to iron out reasoning, just seems too much management work at surface.
Also, i do see crashes of socket.io room getting closed in logs quite often.

**Comment by timolegros** [comment]
1. I have no idea why `/viewSubscriptions` would be called every screen. It is not necessary for subscribing to rooms. Once subscriptions are fetched once they should be locally cached and used to subscribe to rooms just like every other route we use.  We should only subscribe to rooms on hard refresh. Currently, ONLY chain-event subscriptions have rooms. We don't create rooms for any other notification type.
2. Rooms help with fanning out chain-event notifications VIA WEBSOCKETS since that is the only notification type that currently uses Websockets. Websockets have nothing to do with emails so not sure why you're talking about emails.
3. Again, forum notifications aren't sent through Websockets so no we don't have rooms by `object_id` other than chain-events in which case the room_id is the `chain_id`.
4. We can investigate room crashes but then again socket.io manages all those and will spin the room back up if necessary - feel free to create an issue.

The point is, you really should not be concerned about WebSockets. They are at the bottom of the priority list when it comes to notifications in general, at least until we have an actual functioning system and not the dilapidated system we have now. As I said, `/viewSubscriptions` is by far a better thing to focus on. If the route is called too many times and it is very slow then that's a great opportunity for a fix.

**Comment by NakulManchanda** [comment]
Only chain_id and chain events that sounds manageable! Yeah, viewSubscription we should find out why do we call it even once, except screen where we list subscriptions it should not be required.

**Comment by NakulManchanda** [comment]
There are currently four open issues which have preliminary Pull Requests (PRs) attached:

1. [Issue #3438](https://github.com/hicommonwealth/commonwealth/issues/3438): Insert Notification
   - This issue involves an active review that occurred a few weeks ago.
   - We have the option to eliminate the offset column if preferred.
   - It demonstrates superior performance compared to the existing system, as tested via a load test.
   - The PR attached contains actual performance metrics.

2. [Issue #4033](https://github.com/hicommonwealth/commonwealth/issues/4033): Delete Subscription
   - This bug was reported by Rollbar and Tim.
   - It fixes a cascade delete problem.

3. [Issue #4089](https://github.com/hicommonwealth/commonwealth/issues/4089) (Draft): View Subscriptions
   - The goal is to reduce the amount of data retrieved during viewSubs.
   - For example, in extreme cases, as much as 11.2MB of subscription data has been extracted.
   - Currently, complete rows from comments, threads, etc. are being pulled, which poses a risk of running out of memory (OOM) and is unnecessary.
   - Large text rows, such as those in the comment and thread text columns, are some of the main culprits.

4. [Issue #4201](https://github.com/hicommonwealth/commonwealth/issues/4201) (Draft): Obtain comment count, thread count, etc.
   - WIP
   - This initiative relies on Notifications to maintain aggregates like likes, reaction, and comment count.
   - We can utilize the emitNotification function in Notifications to keep track of these counts.
   - Alternatively, this task can be executed as a trigger in the database.

Additionally, a new Load testing tool is available ([PR #4403](https://github.com/hicommonwealth/commonwealth/pull/4403)), which could be leveraged in documenting performance improvements.

**Comment by timolegros** [comment]
No longer relevant.
